test/src/tests/acceptancetests/paginggrouptest/paging_test.go

```diff
@@ -29,209 +29,175 @@ func getPagingClient() PagingClient {
 		c.Assert(err, chk.IsNil)
-		res = resNext
+		c.Assert(page.Values(), chk.NotNil)
+		count++
 	}
 	c.Assert(count, chk.Equals, 10)
 
-	resChan, errChan := pagingClient.GetMultiplePagesRetryFirstComplete(context.Background())
 	count = 0
-	for item := range resChan {
+	for iter, err := pagingClient.GetMultiplePagesRetryFirstAll(context.Background()); iter.NotDone(); err = iter.Next() {
+		c.Assert(err, chk.IsNil)
+		c.Assert(iter.Value().Properties, chk.NotNil)
 		count++
-		c.Assert(item, chk.NotNil)
 	}
 	c.Assert(count, chk.Equals, 10)
-	c.Assert(<-errChan, chk.IsNil)
 }
 
 func (s *PagingGroupSuite) TestGetMultiplePagesRetrySecond(c *chk.C) {
-	res, err := pagingClient.GetMultiplePagesRetrySecond(context.Background())
-	c.Assert(err, chk.IsNil)
-	count := 1
-	for res.NextLink != nil {
-		count++
-		resNext, err := pagingClient.GetMultiplePagesRetrySecondNextResults(res)
+	count := 0
+	for page, err := pagingClient.GetMultiplePagesRetrySecond(context.Background()); page.NotDone(); err = page.Next() {
 		c.Assert(err, chk.IsNil)
-		res = resNext
+		c.Assert(page.Values(), chk.NotNil)
+		count++
 	}
 	c.Assert(count, chk.Equals, 10)
 
-	resChan, errChan := pagingClient.GetMultiplePagesRetrySecondComplete(context.Background())
 	count = 0
-	for item := range resChan {
+	for iter, err := pagingClient.GetMultiplePagesRetrySecondAll(context.Background()); iter.NotDone(); err = iter.Next() {
+		c.Assert(err, chk.IsNil)
+		c.Assert(iter.Value().Properties, chk.NotNil)
 		count++
-		c.Assert(item, chk.NotNil)
 	}
 	c.Assert(count, chk.Equals, 10)
-	c.Assert(<-errChan, chk.IsNil)
 }
 
 func (s *PagingGroupSuite) TestGetSinglePagesFailure(c *chk.C) {
-	res, err := pagingClient.GetSinglePagesFailure(context.Background())
+	page, err := pagingClient.GetSinglePagesFailure(context.Background())
 	c.Assert(err, chk.NotNil)
-	c.Assert(res.StatusCode, chk.Equals, http.StatusBadRequest)
+	c.Assert(page.Response().StatusCode, chk.Equals, http.StatusBadRequest)
 
-	resChan, errChan := pagingClient.GetSinglePagesFailureComplete(context.Background())
 	count := 0
-	for item := range resChan {
+	for iter, err := pagingClient.GetSinglePagesFailureAll(context.Background()); iter.NotDone(); err = iter.Next() {
+		c.Assert(err, chk.IsNil)
+		c.Assert(iter.Value().Properties, chk.NotNil)
 		count++
-		c.Assert(item, chk.NotNil)
 	}
 	c.Assert(count, chk.Equals, 0)
-	c.Assert(<-errChan, chk.NotNil)
 }
 
 func (s *PagingGroupSuite) TestGetMultiplePagesFailure(c *chk.C) {
-	res, err := pagingClient.GetMultiplePagesFailure(context.Background())
+	page, err := pagingClient.GetMultiplePagesFailure(context.Background())
 	c.Assert(err, chk.IsNil)
-	c.Assert(res.NextLink, chk.NotNil)
-	res, err = pagingClient.GetMultiplePagesFailureNextResults(res)
+	c.Assert(page.NotDone(), chk.Equals, true)
+	err = page.Next()
 	c.Assert(err, chk.NotNil)
-	c.Assert(res.StatusCode, chk.Equals, http.StatusBadRequest)
+	//c.Assert(page.Response().StatusCode, chk.Equals, http.StatusBadRequest)
 
```

Comment:
Is this not working as intended?

Reply:
No, the page struct is only updated on success else you wouldn't be able to try again (e.g. transient failure case).  The right way to fix this is to unify error reporting so that one can examine the non-nil error object to get the status code.

Reply:
Maybe the right thing to do is update the raw response inside  this struct, regardless if operation was successful?

Reply:
But now the page type is in a half-baked state which I don't think is very good.  We really should fix how we report errors so that they are more than just error message strings (we've had several complaints about this and I believe there's an issue open for it now).

Reply:
Will update the test to get the status code from the returned error.

Reply:
Fixed.